### PR TITLE
Preprocess an entire folder of .json or .json.gz files into a single .bin and .idx file.

### PR DESCRIPTION
**Pull request by MaximumEntropy**

### CODE REVIEW

File: scripts/nlp_language_modeling/preprocess_data_for_megatron.py [hunk 1]
```diff
@@ -18,9 +18,11 @@
 import gzip
 import json
 import multiprocessing
+import os
 import sys
 import time
 
+import ftfy
```
Comment (ericharper) [inline scripts/nlp_language_modeling/preprocess_data_for_megatron.py:25]: Do we need to add this to requirements_nlp.txt?

File: scripts/nlp_language_modeling/preprocess_data_for_megatron.py [hunk 2]
```diff
@@ -132,6 +136,12 @@ def get_args():
     group = parser.add_argument_group(title='runtime')
     group.add_argument('--workers', type=int, default=1, help='Number of worker processes to launch')
     group.add_argument('--log-interval', type=int, default=100, help='Interval between progress updates')
+    group.add_argument(
+        '--preproc-folder',
+        action='store_true',
+        help='If set, will preprocess all .json or .json.gz files into a single .bin and .idx file. Folder path provided via the --input arg',
```
Comment (ericharper) [inline scripts/nlp_language_modeling/preprocess_data_for_megatron.py:142]: Can you add help text to the --input arg explaining that input can be a single file or folder?

File: scripts/nlp_language_modeling/preprocess_data_for_megatron.py [hunk 3]
```diff
@@ -152,12 +162,19 @@ def get_args():
...
+        json_files = [
+            os.path.join(args.input, f) for f in files_in_folder if f.endswith('.json') or f.endswith('.json.gz')
+        ]
+        if len(json_files) == 0:
+            raise FileNotFoundError('No .json or .json.gz files found in folder.')
+        else:
+            print(f'Found {len(json_files)} .json or .json.gz files.')
     else:
-        fin = open(args.input, 'r', encoding='utf-8')
+        json_files = [args.input]
```
Comment (ericharper) [inline scripts/nlp_language_modeling/preprocess_data_for_megatron.py:177]: might as well check the path exists here?

**Review by ericharper — CHANGES_REQUESTED** [review]
Great idea! Just a few small comments.

File: scripts/nlp_language_modeling/preprocess_data_for_megatron.py (hunk 1 above)
Comment (MaximumEntropy) [reply scripts/nlp_language_modeling/preprocess_data_for_megatron.py:25]: Done!

File: scripts/nlp_language_modeling/preprocess_data_for_megatron.py (hunk 2 above)
Comment (MaximumEntropy) [reply scripts/nlp_language_modeling/preprocess_data_for_megatron.py:142]: Good point! Done.

File: scripts/nlp_language_modeling/preprocess_data_for_megatron.py (hunk 3 above)
Comment (MaximumEntropy) [reply scripts/nlp_language_modeling/preprocess_data_for_megatron.py:178]: Added a check here as well. Good catch.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**